server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -1000,30 +995,32 @@ private boolean tryRelocateShard(ModelNode minNode, ModelNode maxNode, String id
                 }
                 ShardRouting candidate = null;
                 final AllocationDeciders deciders = allocation.deciders();
+                final List<ShardRouting> shardRoutings = new ArrayList<>(index.numShards());
                 for (ShardRouting shard : index) {
                     if (shard.started()) {
                         // skip initializing, unassigned and relocating shards we can't relocate them anyway
-                        Decision allocationDecision = deciders.canAllocate(shard, minNode.getRoutingNode(), allocation);
-                        Decision rebalanceDecision = deciders.canRebalance(shard, allocation);
-                        if (((allocationDecision.type() == Type.YES) || (allocationDecision.type() == Type.THROTTLE))
-                                && ((rebalanceDecision.type() == Type.YES) || (rebalanceDecision.type() == Type.THROTTLE))) {
-                            if (maxNode.containsShard(shard)) {
-                                // simulate moving shard from maxNode to minNode
-                                final float delta = weight.weightShardAdded(
-                                    this, minNode, idx) - weight.weightShardRemoved(this, maxNode, idx);
-                                if (delta < minCost ||
-                                        (candidate != null && Float.compare(delta, minCost) == 0 && candidate.id() > shard.id())) {
-                                    /* this last line is a tie-breaker to make the shard allocation alg deterministic
-                                     * otherwise we rely on the iteration order of the index.getAllShards() which is a set.*/
-                                    minCost = delta;
-                                    candidate = shard;
-                                    decision = new Decision.Multi().add(allocationDecision).add(rebalanceDecision);
-                                }
-                            }
+                        if (maxNode.containsShard(shard)) {
+                            shardRoutings.add(shard);
                         }
                     }
                 }
 
+                // look for a relocation candidate, in descending order of shard id so that the decision is deterministic
+                shardRoutings.sort(BY_DESCENDING_SHARD_ID);
+                for (ShardRouting shard : shardRoutings) {
+                    final Decision rebalanceDecision = deciders.canRebalance(shard, allocation);
+                    if ((rebalanceDecision.type() != Type.YES) && (rebalanceDecision.type() != Type.THROTTLE)) {
+                        continue;
```

Comment:
perhaps just
```
if (rebalanceDecision.type() == Type.NO) {
  continue;
}
```

and the same for the allocationDecision below

Reply:
👍 b4f90c20f3c439f2c7d614767bd77866e86f8d17